Sources/NIOHTTPCompression/HTTPResponseDecompressor.swift

```diff
@@ -0,0 +1,207 @@
+            switch self {
+            case .deflate:
+                return 15
+            case .gzip:
+                return 15 + 16
+            }
+        }
+    }
+
+    private enum State {
+        case empty
+        case compressed(CompressionAlgorithm, Int)
+    }
+
+    private let limit: DecompressionLimit
+    private var state = State.empty
+    private var stream = z_stream()
+    private var inflated = 0
+
+    public init(limit: DecompressionLimit) {
+        self.limit = limit
+    }
+
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        let request = self.unwrapOutboundIn(data)
+        switch request {
+        case .head(var head):
+            if head.headers.contains(name: "Accept-Encoding") {
+                context.write(data, promise: promise)
+            } else {
+                head.headers.replaceOrAdd(name: "Accept-Encoding", value: "deflate, gzip")
+                context.write(self.wrapOutboundOut(.head(head)), promise: promise)
+            }
+        default:
+            context.write(data, promise: promise)
+        }
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        switch self.unwrapInboundIn(data) {
+        case .head(let head):
+            let contentType = head.headers[canonicalForm: "Content-Encoding"].first?.lowercased()
+            let algorithm = CompressionAlgorithm(header: contentType)
+
+            let length = head.headers[canonicalForm: "Content-Length"].first.flatMap { Int($0) }
+
+            if let algorithm = algorithm, let length = length {
+                do {
+                    try self.initializeDecoder(encoding: algorithm, length: length)
+                } catch {
+                    context.fireErrorCaught(error)
+                    return
+                }
+            }
+
+            context.fireChannelRead(data)
+        case .body(var part):
+            switch self.state {
+            case .compressed(_, let originalLength):
+                while part.readableBytes > 0 {
+                    do {
+                        var buffer = context.channel.allocator.buffer(capacity: 16384)
+                        try self.stream.inflatePart(input: &part, output: &buffer)
+                        self.inflated += buffer.readableBytes
+
+                        if self.limit.exceeded(compressed: originalLength, decompressed: self.inflated) {
+                            context.fireErrorCaught(DecompressionError.limit)
+                            return
+                        }
+
+                        context.fireChannelRead(self.wrapInboundOut(.body(buffer)))
+                    } catch {
+                        context.fireErrorCaught(error)
+                        return
+                    }
+                }
+            default:
+                context.fireChannelRead(data)
+            }
+        case .end:
+            deflateEnd(&self.stream)
```

Comment:
This should be `inflateEnd`. Presumably also this shouldn't be unconditionally called, but only called when we have actually initialised the decoder?

Reply:
good catch, fixed, thanks!